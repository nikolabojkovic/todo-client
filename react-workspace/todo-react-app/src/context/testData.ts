import { 
  ITodo,
  IPaging, 
  ISort, 
  SortDirection, 
  IFilter,
  StateFilter, 
  IGeneralSettings, 
  IPaginationSettings, 
  ISearchSettings, 
  ISettings, 
  ListContainerType, 
  PaginationType, 
  BackgroundColor,
  IThemeSettings,
  BsThemes
} from "../models";
import { DisplayMode, IState } from "./";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  displayMode: DisplayMode.All,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings,
    theme: {
      backgroundColor: BackgroundColor.DarkGray,
      primaryColor: '#ff9900',
      bsTheme: BsThemes.Light,
      primaryColorTopCord: -35,
      primaryColorLefCord: 81
    } as IThemeSettings
  } as ISettings
} as IState;