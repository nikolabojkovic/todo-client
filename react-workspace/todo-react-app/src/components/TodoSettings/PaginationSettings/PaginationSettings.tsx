import { Dropdown, Stack } from "react-bootstrap";

import { useTodoList, useTodoListDispatch } from "../../../context";
import { PaginationType, IAction, TodoActions } from "../../../models";

export function PaginationSettings() {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  const ifPaginationDisabled = !todoList.settings.general.isPaginationEnabled;
  
  function handlePaginationTypeSelect(option: PaginationType) {
    dispatch({
      type: TodoActions.settingsUpdated,
      payload: {
        ...todoList.settings,
        pagination: {
          ...todoList.settings.pagination,
          paginationType: option
        }
      }
    } as IAction);
  }

  function handleVisiblePagesSelect(option: number) {
    dispatch({
      type: TodoActions.settingsUpdated,
      payload: {
        ...todoList.settings,
        pagination: {
          ...todoList.settings.pagination,
          maxVisiblePages: option
        }
      }
    } as IAction);
  }

  return (
    <div className={ "App__settings__group" + (ifPaginationDisabled ? " App__settings__group--disabled" : "") }>
      <label className={"App__settings__group-label" + (ifPaginationDisabled ? " App__settings__group-label--disabled" : "") }>
        Pagination
      </label>
      <div className={ "App__settings__group__item" + (ifPaginationDisabled ? " App__settings__group__item--disabled" : "") }>
        <Stack direction="horizontal" gap={2}>
          <label>Pagination type</label>
          <Dropdown className="ms-auto">
            <Dropdown.Toggle
              data-testid="dropdown-pagination-type"
              variant="outline-secondary"
              className="action-button ps-2 pe-2"
              id="dropdown-pagination-type"
              disabled={ifPaginationDisabled}
            >
              {' '}{todoList.settings.pagination.paginationType}{' '}
            </Dropdown.Toggle>
            <Dropdown.Menu>
              <Dropdown.Item 
                data-testid="pagination-type-option-classic" 
                onClick={() => handlePaginationTypeSelect(PaginationType.Classic)}
              >
                { PaginationType.Classic }
              </Dropdown.Item>
              <Dropdown.Item 
                data-testid="pagination-type-option-rotate" 
                onClick={() => handlePaginationTypeSelect(PaginationType.Rotate)}
              >
                { PaginationType.Rotate }
              </Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </Stack>
      </div>
      <div className={ "App__settings__group__item" + (ifPaginationDisabled ? " App__settings__group__item--disabled" : "") }>
        <Stack direction="horizontal" gap={2}>
          <label>Visible pages in navigation group</label>
          <Dropdown className="ms-auto">
            <Dropdown.Toggle
              data-testid="dropdown-visible-pages-number"
              variant="outline-secondary"
              className="action-button ps-2 pe-2"
              id="dropdown-visible-pages-number"
              disabled={ifPaginationDisabled}
            >
              {' '}{todoList.settings.pagination.maxVisiblePages}{' '}
            </Dropdown.Toggle>
            <Dropdown.Menu>
              <Dropdown.Item data-testid="visible-pages-option-1" onClick={() => handleVisiblePagesSelect(1)}>1</Dropdown.Item>
              <Dropdown.Item data-testid="visible-pages-option-2" onClick={() => handleVisiblePagesSelect(2)}>2</Dropdown.Item>
              <Dropdown.Item data-testid="visible-pages-option-3" onClick={() => handleVisiblePagesSelect(3)}>3</Dropdown.Item>
              <Dropdown.Item data-testid="visible-pages-option-4" onClick={() => handleVisiblePagesSelect(4)}>4</Dropdown.Item>
              <Dropdown.Item data-testid="visible-pages-option-5" onClick={() => handleVisiblePagesSelect(5)}>5</Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </Stack>
      </div>
    </div>
  );
}