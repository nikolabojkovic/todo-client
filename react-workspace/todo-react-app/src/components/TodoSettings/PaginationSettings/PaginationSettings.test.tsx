import renderer, { act } from 'react-test-renderer';
import { fireEvent, render, screen  } from '@testing-library/react';

import { IState, TodosContext, TodosDispatchContext, stateTestData } from "../../../context";
import { PaginationSettings } from "./PaginationSettings";
import { IAction, IGeneralSettings, IPaginationSettings, ISettings, PaginationType, TodoActions } from '../../../models';
import userEvent from '@testing-library/user-event';

describe('PaginationSettings', () => {
  const globalContext = {
    state: {
      ...stateTestData,
    },
    dispatch: jest.fn()
  };
  
  it('match snapshot pagination settings', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false 
      }
    };
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
          <PaginationSettings></PaginationSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot pagination settings disabled', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            isPaginationEnabled: false
          } as IGeneralSettings
        } as ISettings
      } as IState
    };
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
          <PaginationSettings></PaginationSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  describe('actions', () => {
    it('should update settings on pagination type classic option selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              isPaginationEnabled: true
            } as IGeneralSettings,
            pagination: {
              ...globalContext.state.settings.pagination,
              paginationType: PaginationType.Rotate
            } as IPaginationSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <PaginationSettings></PaginationSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const paginationTypeDropdown = screen.getByTestId('dropdown-pagination-type');
  
      await act(async () => {      
        return await userEvent.click(paginationTypeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          pagination: {
            ...context.state.settings.pagination,
            paginationType: PaginationType.Classic
          } as IPaginationSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('pagination-type-option-classic');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });
    it('should update settings on pagination type rotate option selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              isPaginationEnabled: true
            } as IGeneralSettings,
            pagination: {
              ...globalContext.state.settings.pagination,
              paginationType: PaginationType.Classic
            } as IPaginationSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <PaginationSettings></PaginationSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const paginationTypeDropdown = screen.getByTestId('dropdown-pagination-type');
  
      await act(async () => {      
        return await userEvent.click(paginationTypeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          pagination: {
            ...context.state.settings.pagination,
            paginationType: PaginationType.Rotate
          } as IPaginationSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('pagination-type-option-rotate');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });

    it('should update settings on pagination visible pages option 1 selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              isPaginationEnabled: true
            } as IGeneralSettings,
            pagination: {
              ...globalContext.state.settings.pagination,
              paginationType: PaginationType.Classic,
              maxVisiblePages: 3
            } as IPaginationSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <PaginationSettings></PaginationSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const paginationTypeDropdown = screen.getByTestId('dropdown-visible-pages-number');
  
      await act(async () => {      
        return await userEvent.click(paginationTypeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          pagination: {
            ...context.state.settings.pagination,
            paginationType: PaginationType.Classic,
            maxVisiblePages: 1
          } as IPaginationSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('visible-pages-option-1');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });
    it('should update settings on pagination visible pages option 2 selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              isPaginationEnabled: true
            } as IGeneralSettings,
            pagination: {
              ...globalContext.state.settings.pagination,
              paginationType: PaginationType.Classic,
              maxVisiblePages: 3
            } as IPaginationSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <PaginationSettings></PaginationSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const paginationTypeDropdown = screen.getByTestId('dropdown-visible-pages-number');
  
      await act(async () => {      
        return await userEvent.click(paginationTypeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          pagination: {
            ...context.state.settings.pagination,
            paginationType: PaginationType.Classic,
            maxVisiblePages: 2
          } as IPaginationSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('visible-pages-option-2');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });
    it('should update settings on pagination visible pages option 3 selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              isPaginationEnabled: true
            } as IGeneralSettings,
            pagination: {
              ...globalContext.state.settings.pagination,
              paginationType: PaginationType.Classic,
              maxVisiblePages: 5
            } as IPaginationSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <PaginationSettings></PaginationSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const paginationTypeDropdown = screen.getByTestId('dropdown-visible-pages-number');
  
      await act(async () => {      
        return await userEvent.click(paginationTypeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          pagination: {
            ...context.state.settings.pagination,
            paginationType: PaginationType.Classic,
            maxVisiblePages: 3
          } as IPaginationSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('visible-pages-option-3');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });
    it('should update settings on pagination visible pages option 4 selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              isPaginationEnabled: true
            } as IGeneralSettings,
            pagination: {
              ...globalContext.state.settings.pagination,
              paginationType: PaginationType.Classic,
              maxVisiblePages: 3
            } as IPaginationSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <PaginationSettings></PaginationSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const paginationTypeDropdown = screen.getByTestId('dropdown-visible-pages-number');
  
      await act(async () => {      
        return await userEvent.click(paginationTypeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          pagination: {
            ...context.state.settings.pagination,
            paginationType: PaginationType.Classic,
            maxVisiblePages: 4
          } as IPaginationSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('visible-pages-option-4');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });
    it('should update settings on pagination visible pages option 5 selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              isPaginationEnabled: true
            } as IGeneralSettings,
            pagination: {
              ...globalContext.state.settings.pagination,
              paginationType: PaginationType.Classic,
              maxVisiblePages: 3
            } as IPaginationSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <PaginationSettings></PaginationSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const paginationTypeDropdown = screen.getByTestId('dropdown-visible-pages-number');
  
      await act(async () => {      
        return await userEvent.click(paginationTypeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          pagination: {
            ...context.state.settings.pagination,
            paginationType: PaginationType.Classic,
            maxVisiblePages: 5
          } as IPaginationSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('visible-pages-option-5');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });
   });
});