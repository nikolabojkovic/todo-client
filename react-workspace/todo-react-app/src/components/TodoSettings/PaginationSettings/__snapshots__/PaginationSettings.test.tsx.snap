// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationSettings match snapshot pagination settings 1`] = `
<div
  className="App__settings__group"
>
  <label
    className="App__settings__group-label"
  >
    Pagination
  </label>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Pagination type
      </label>
      <div
        className="ms-auto dropdown"
      >
        <button
          aria-expanded={false}
          className="action-button ps-2 pe-2 dropdown-toggle btn btn-outline-secondary"
          data-testid="dropdown-pagination-type"
          disabled={false}
          id="dropdown-pagination-type"
          onClick={[Function]}
          type="button"
        >
           
          Classic
           
        </button>
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Visible pages in navigation group
      </label>
      <div
        className="ms-auto dropdown"
      >
        <button
          aria-expanded={false}
          className="action-button ps-2 pe-2 dropdown-toggle btn btn-outline-secondary"
          data-testid="dropdown-visible-pages-number"
          disabled={false}
          id="dropdown-visible-pages-number"
          onClick={[Function]}
          type="button"
        >
           
          3
           
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaginationSettings match snapshot pagination settings disabled 1`] = `
<div
  className="App__settings__group App__settings__group--disabled"
>
  <label
    className="App__settings__group-label App__settings__group-label--disabled"
  >
    Pagination
  </label>
  <div
    className="App__settings__group__item App__settings__group__item--disabled"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Pagination type
      </label>
      <div
        className="ms-auto dropdown"
      >
        <button
          aria-expanded={false}
          className="action-button ps-2 pe-2 dropdown-toggle btn btn-outline-secondary"
          data-testid="dropdown-pagination-type"
          disabled={true}
          id="dropdown-pagination-type"
          onClick={[Function]}
          type="button"
        >
           
          Classic
           
        </button>
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item App__settings__group__item--disabled"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Visible pages in navigation group
      </label>
      <div
        className="ms-auto dropdown"
      >
        <button
          aria-expanded={false}
          className="action-button ps-2 pe-2 dropdown-toggle btn btn-outline-secondary"
          data-testid="dropdown-visible-pages-number"
          disabled={true}
          id="dropdown-visible-pages-number"
          onClick={[Function]}
          type="button"
        >
           
          3
           
        </button>
      </div>
    </div>
  </div>
</div>
`;
